src/tests/unit/tests/common/message-creators/bug-action-message-creator.test.ts

```diff
@@ -1,77 +1,70 @@
 import { Messages } from '../../../../../common/messages';
 import { TelemetryDataFactory } from '../../../../../common/telemetry-data-factory';
 import {
+    BaseTelemetryData,
     FILE_ISSUE_CLICK,
     FileIssueClickService,
-    FileIssueClickTelemetryData,
-    SettingsOpenTelemetryData,
+    SettingsOpenSourceItem,
     TelemetryEventSource,
+    TriggeredBy,
 } from '../../../../../common/telemetry-events';
 import { EventStubFactory } from '../../../common/event-stub-factory';
 
 describe('BugActionMessageCreator', () => {
     const source: TelemetryEventSource = TelemetryEventSource.TargetPage;
-    const eventStubFactory = new EventStubFactory();
-    const event = eventStubFactory.createMouseClickEvent() as any;
+    const eventStub = new EventStubFactory().createKeypressEvent() as any;
+    const telemetryStub: BaseTelemetryData = {
+        triggeredBy: 'test' as TriggeredBy,
+        source,
+    };
+
+    let telemetryFactoryMock: IMock<TelemetryDataFactory>;
+    let dispatcherMock: IMock<ActionMessageDispatcher>;
+
     let testSubject: BugActionMessageCreator;
-    let postMessageMock: IMock<(msg: any) => void>;
-    let tabId: number;
 
     beforeEach(() => {
-        tabId = -1;
-        postMessageMock = Mock.ofInstance(message => {});
-        testSubject = new BugActionMessageCreator(postMessageMock.object, tabId, new TelemetryDataFactory(), source);
+        telemetryFactoryMock = Mock.ofType<TelemetryDataFactory>();
+        dispatcherMock = Mock.ofType<ActionMessageDispatcher>();
+
+        testSubject = new BugActionMessageCreator(dispatcherMock.object, telemetryFactoryMock.object, source);
     });
 
-    test('openSettingsPanel', () => {
-        const telemetry: SettingsOpenTelemetryData = {
-            triggeredBy: 'mouseclick',
-            source,
-            sourceItem: 'fileIssueSettingsPrompt',
-        };
-        const payload: BaseActionPayload = {
-            telemetry,
-        };
-        const expectedMessage: Message = {
-            messageType: Messages.SettingsPanel.OpenPanel,
-            tabId,
-            payload,
-        };
+    it('dispatch message for openSettingsPanel', () => {
+        const telemetry = { ...telemetryStub, sourceItem: 'sourceItem' as SettingsOpenSourceItem };
+        telemetryFactoryMock
+            .setup(factory => factory.forSettingsPanelOpen(eventStub, source, 'fileIssueSettingsPrompt'))
+            .returns(() => telemetry);
 
-        postMessageMock.setup(pm => pm(It.isValue(expectedMessage))).verifiable();
+        dispatcherMock.setup(dispatcher =>
+            dispatcher.dispatchMessage({
+                messageType: Messages.SettingsPanel.OpenPanel,
+                payload: {
+                    telemetry,
+                },
+            }),
+        );
 
-        testSubject.openSettingsPanel(event);
+        testSubject.openSettingsPanel(eventStub);
```

Comment:
If I comment out this line, the test still passes. Need a `.verifiable(Times.once())` on the mock setup.

Reply:
(The `telemetryFactoryMock` is also not verified, but I can see that as an implementation detail of the test and not part of the contract being examined.)

Reply:
Good catch on the missing `verifiable`
As per the `telemetryFactoryMock`, in general, use of `verifiable` is not needed when the return value of the mock alter the result of the tested function. In this case, verifying the dispatcher is proof enough the telemetry factory was used indeed.